File: 701/701.cpp
/*!***********************
* @brief:  leetcode-701-二叉搜索树中的插入操作
* @date:   2019/5/8
* @envir:  gcc 7.4.0
* @author: luhao
* ***********************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* insertIntoBST(TreeNode* root, int val) {
        TreeNode *p = root;
        TreeNode *z = NULL;
        while (p!=NULL){
            z = p;
            if (val<p->val)
                p = p->left;
            else
                p = p->right;
        }
        if (val < z->val) //插在z左边
            z->left = new TreeNode(val);
        else //插在z右边
            z->right = new TreeNode(val);
        return root;
    }
};

File: 701/701.py
# @brief:  leetcode-701-二叉搜索树中的插入操作
# @envir:  python 2.7
# @date:   2019/5/7
# @author: luhao

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def insertIntoBST(self, root, val):
        """
        :type root: TreeNode
        :type val: int
        :rtype: TreeNode
        """
        p = root
        x = None
        while p!=None:
            x = p
            if val < p.val:
                p = p.left
            else:
                p = p.right
        
        if val < x.val:
            x.left = TreeNode(val)
        else:
            x.right = TreeNode(val)
        return root
        
